apps/omg/lib/omg/state/core.ex

```diff
@@ -161,16 +183,25 @@ defmodule OMG.State.Core do
   """
   @spec exec(state :: t(), tx :: Transaction.Recovered.t(), fees :: Fees.optional_fee_t()) ::
           {:ok, {Transaction.tx_hash(), pos_integer, non_neg_integer}, t()}
-          | {{:error, Validator.exec_error()}, t()}
+          | {{:error, Validator.process_error()}, t()}
   def exec(%Core{} = state, %Transaction.Recovered{} = tx, fees) do
     tx_hash = Transaction.raw_txhash(tx)
 
-    case Validator.can_apply_spend(state, tx, fees) do
-      true ->
+    case Validator.can_process(state, tx, fees) do
+      {:ok, :apply_spend, fees_paid} ->
+        {:ok, {tx_hash, state.height, state.tx_index},
+         state
+         |> apply_spend(tx)
+         |> add_pending_tx(tx)
+         |> collect_fees(fees_paid)}
+
+      {:ok, :claim_fees, claimed_token} ->
         {:ok, {tx_hash, state.height, state.tx_index},
          state
          |> apply_spend(tx)
-         |> add_pending_tx(tx)}
+         |> add_pending_tx(tx)
+         |> claim_token(claimed_token |> Map.keys() |> hd())
```

Comment:
so this `keys |> hd` looks like aftermath from the overloading of `Transaction.Protocol.can_apply?`.

Having thought about this a bit, I think you should abide by what `Transaction.Protocol.can_apply?` originally required the implementors to return, being `{:ok, fees paid per token}`. This is an abstract thing and let's not touch it, i.e. return 0 for `FeeClaimTransaction`s

Now to be able to do what you want, consider putting the fee-claiming pipeline outside of this `case Validator.can_process` pipeline.

So:
1. the `Validator` told you to `apply_spend` and `add_pending_tx` (which now happens uniformly for all tx types, so no need to branch _here_
2. if this happened to be a Fee transaction then do these two new steps (`claim_token` & `disallow_payments`)

In other words, handle the fee transactions as a post-processing step instead of bloating the `Transaction.Protocol` abstraction